Simplified and corrected logic around context cancelation in refactored QueueSet

MikeSpreitzer opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change

/kind bug

> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
You can view this as extending #87350 or rebasing #87059 .  This PR fixes a bugs that were in QueueSet before it was refactored and were retained in the refactoring.  There are two commits here: the first is a copy of #87350 (the QueueSet refactor) and the second is the bug fix.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

MikeSpreitzer committed:
Simplified logic around context cancel, removing bugs

Previously, a `decisionCancel` could overwrite a `decisionReject` or
`decisionExecute`, causing confusion.  Now a request gets exactly one
decision and there is no confusion.

Also added write-once to the promise package and refactored.

MikeSpreitzer committed:
Refactored QueueSet configuration into two phases

So that errors can be detected before resolving concurrency shares
into concurrency counts.

MikeSpreitzer commented:
/sig api-machinery

MikeSpreitzer commented:
/cc @lavalamp 
/cc @deads2k 
@bjrara 
/cc @jktomer 
/cc @tedyu

MikeSpreitzer commented:
The force-push to 1c092bf6359 is a rebase onto master.

fedebongio commented:
/assign @lavalamp @jktomer

MikeSpreitzer commented:
@jktomer

MikeSpreitzer committed:
Tweak new names

lavalamp reviewed (commented):
Thanks for adding the test...

One last thought, do we actually use the WriteMultiple types? Maybe leave them out if not, since non-use implies we won't find the bugs? (Although I guess if you follow the suggestion to combine most of the implementation, there's not much chance of a bug in the remaining ~2 unique lines...)

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/promise/interface.go:
```diff
@@ -16,27 +16,87 @@ limitations under the License.
+type LockingReadable interface {
+	Readable
+
+	// GetLocked is like Get but the caller must already hold the lock
```
For a LockingReadable, the Readable methods acquire the mutex.  I will add commentary.

MikeSpreitzer replied:
done

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/promise/interface.go:
```diff
@@ -16,27 +16,87 @@ limitations under the License.
+	// Set normally writes a value into this variable, unblocks every
+	// goroutine waiting for this variable to have a value, and
+	// returns true.  In the unhappy case that this variable is
+	// already set, this method returns false.
```
OK, will say so.

MikeSpreitzer replied:
done

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/promise/interface.go:
```diff
@@ -16,27 +16,87 @@ limitations under the License.
+	WriteOnceOnly
+
+	// SetLocked is like Set but the caller must already hold the lock
+	SetLocked(interface{}) bool
```
I did not need it.  But, now that you mention it, doing that would make this easier to extend.

MikeSpreitzer replied:
To be clear, there would be fewer interfaces here if golang allowed diamond-shaped inheritence among interfaces.  But it does not; see https://play.golang.org/p/Eh7l2gQzpSG for a demo.

MikeSpreitzer replied:
done

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/promise/lockingpromise/lockingpromise.go:
```diff
@@ -23,57 +23,108 @@ import (
-		lp.waitingCount++
-		lp.activeCounter.Add(-1)
-		lp.cond.Wait()
+func (wr *writeMultiple) SetLocked(value interface{}) {
```
Right, the shared stuff can be moved to promisoid.

MikeSpreitzer replied:
Since I recently changed my mind about which one I need, that could plausibly happen again soon.  Once the client code settles down I might be more interested in trimming.  But you're right, I think it's not a big deal.

MikeSpreitzer replied:
moved shared SetLocked behavior to promisoid.

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/promise/lockingpromise/lockingpromise_test.go:
```diff
@@ -43,8 +43,74 @@ func TestLockingPromise(t *testing.T) {
+	aval := &now
+	wr.Set(aval)
+	clock.Run(nil)
+	time.Sleep(time.Second)
```
No, `clock.Run` takes care of this if the code is correct.  Before every attempt to advance the fake time, it waits for all counted goroutines to finish or block.  The sleep is only there to catch bugs.

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/promise/lockingpromise/lockingpromise_test.go:
```diff
@@ -43,8 +43,74 @@ func TestLockingPromise(t *testing.T) {
+	wr.Set(aval)
+	clock.Run(nil)
+	time.Sleep(time.Second)
+	if atomic.LoadInt32(&gots) != 1 {
```
These promisoids are explicitly intended to participate in the goroutine counting that FakeEventClock uses, so it is proper for the test to exercise that.

MikeSpreitzer committed:
brushed up according to review

Added LockingWriteMultipleOnly and LockingWriteOnceOnly interfaces,
so that further extensions are possible (in this package or others).

Moved common SetLocked behavior into promisoid.

Made comments say things that were implied.

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/promise/interface.go:
```diff
@@ -16,27 +16,87 @@ limitations under the License.
-// or more times (unlike a traditional "promise", which can be written
-// only once).
-type Mutable interface {
+// This file defines interfaces for promsies and futures and related
```
thanks, fixed

MikeSpreitzer commented:
/retest

lavalamp commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/promise/lockingpromise/lockingpromise.go:
```diff
@@ -23,12 +23,12 @@ import (
+// the given counter is decremented for a goroutine waiting for this
+// varible to be set and incremented when such a goroutine is
+// unblocked.
 type promisoid struct {
```
BTW love the name `promisoid`-- very clear and short, but also seems to be newly coined, at least a search showed only typos in bad OCR files.  :)

lavalamp commented:
/approve
/lgtm

/test pull-kubernetes-e2e-gce

MikeSpreitzer commented:
/retest
